### PR TITLE
Upgrades to Spark 3.4/JRE 17 and fixes all high/critical CVEs

### PR DESCRIPTION
**Pull request by codefromthecrypt**
I manually tested this on all three storage types in docker as well (using zipkin's docker/examples instructions)

### CODE REVIEW

**Comment by codefromthecrypt** [comment]
trivy is clean now, so even if we don't release until 8.14 final... at least we can apply the security settings same as other repos after merge:

```
$ trivy repo .
2024-04-15T15:56:48.696-1000	INFO	Need to update DB
2024-04-15T15:56:48.696-1000	INFO	DB Repository: ghcr.io/aquasecurity/trivy-db:2
2024-04-15T15:56:48.696-1000	INFO	Downloading DB...
45.03 MiB / 45.03 MiB [------------------------------------------------------------------------------------------------------------------] 100.00% 2.58 MiB p/s 18s
2024-04-15T15:57:07.461-1000	INFO	Vulnerability scanning is enabled
2024-04-15T15:57:07.461-1000	INFO	Secret scanning is enabled
2024-04-15T15:57:07.461-1000	INFO	If your scanning is slow, please try '--scanners vuln' to disable secret scanning
2024-04-15T15:57:07.462-1000	INFO	Please see also https://aquasecurity.github.io/trivy/v0.50/docs/scanner/secret/#recommendation for faster secret detection
2024-04-15T15:57:07.731-1000	INFO	Number of language-specific files: 5
2024-04-15T15:57:07.731-1000	INFO	Detecting pom vulnerabilities...
```

File: cassandra3/src/main/java/zipkin2/dependencies/cassandra3/CassandraDependenciesJob.java [hunk 1]
```diff
@@ -150,6 +150,9 @@ public CassandraDependenciesJob build() {
     df.setTimeZone(TimeZone.getTimeZone("UTC"));
     this.dateStamp = df.format(new Date(builder.day));
     this.conf = new SparkConf(true).setMaster(builder.sparkMaster).setAppName(getClass().getName());
+    if (builder.sparkMaster.startsWith("local[")) {
+      conf.set("spark.driver.bindAddress", "127.0.0.1");
```
Comment (codefromthecrypt) [inline cassandra3/src/main/java/zipkin2/dependencies/cassandra3/CassandraDependenciesJob.java:154]: this is a spark 3.4 thing

File: elasticsearch/src/main/java/zipkin2/dependencies/elasticsearch/ElasticsearchDependenciesJob.java [hunk 2]
```diff
@@ -167,33 +170,27 @@ private static String getSystemPropertyAsFileResource(String key) {
...
-          .reduceByKey((l, r) -> DependencyLink.newBuilder()
-            .parent(l.parent())
-            .child(l.child())
-            .callCount(l.callCount() + r.callCount())
-            .errorCount(l.errorCount() + r.errorCount())
-            .build())
-          .values()
-          .map(DEPENDENCY_LINK_JSON);
+    JavaRDD<Map<String, Object>> links;
+    try (JavaSparkContext sc = new JavaSparkContext(conf)) {
```
Comment (codefromthecrypt) [inline elasticsearch/src/main/java/zipkin2/dependencies/elasticsearch/ElasticsearchDependenciesJob.java:179]: this is just polish as we can use try/resources with some of the drivers

File: main/src/main/resources/log4j2.properties [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# Set everything to be logged to the console
```
Comment (codefromthecrypt) [inline main/src/main/resources/log4j2.properties:1]: this is also spark 3.4 thing (log4j 2 not 1.2 config)

File: pom.xml [hunk 4]
```diff
@@ -187,6 +224,53 @@
         <type>pom</type>
         <scope>import</scope>
       </dependency>
+
+      <!-- CVE fix versions -->
```
Comment (codefromthecrypt) [inline pom.xml:228]: to keep the build without CVE will be difficult, but anyway at least once it is clean ;)

**Review by codefromthecrypt — COMMENTED** [review]
https://github.com/elastic/elasticsearch-hadoop/issues/2187 unlocked this (releasing in elastic-hadoop 8.14, but not sure when)

other notes below

File: cassandra3/src/main/java/zipkin2/dependencies/cassandra3/CassandraDependenciesJob.java (hunk 1 above)
Comment (codefromthecrypt) [reply cassandra3/src/main/java/zipkin2/dependencies/cassandra3/CassandraDependenciesJob.java:154]: basically it tries to detect with the hostname, which isn't needed for local mode anyway.

**Comment by codefromthecrypt** [comment]
some big wins from merging this include:
* we can switch the bytecode floor of zipkin-storage to 17 (with the other server and collector code)
  * this gets us to a [more supportable version of jooq](https://github.com/openzipkin/zipkin/blob/master/zipkin-storage/mysql-v1/pom.xml#L23-L24) for example.
* we can use the [same security settings as other prod repos](https://github.com/openzipkin/zipkin-reporter-java/pull/267) since the dep tree isn't impossible anymore

**Comment by codefromthecrypt** [comment]
oh yeah spent so much time doing this I forgot why.. I was trying to polish this up prior to adding dependencies to helm. There was a point where I though maybe we need to rewrite the entire thing (like in beam) to solve the revlock. I'm glad it didn't get that far.

https://github.com/openzipkin/zipkin-helm/pull/11

**Review by shakuzen — APPROVED** [review]

**Comment by codefromthecrypt** [comment]
https://github.com/openzipkin/zipkin/pull/3763 for zipkin changes we can now do

**Review by anuraaga — APPROVED** [review]

File: docker/bin/start-zipkin-dependencies [hunk 5]
```diff
@@ -8,4 +8,24 @@
 # args override it (Ex. 2020-10-31)
 set -eu
 
-exec java ${JAVA_OPTS} -Djava.io.tmpdir=/tmp -cp classes zipkin2.dependencies.ZipkinDependenciesJob $@
+# Spark 3.4 module config from:
+# https://github.com/apache/spark/blob/branch-3.4/launcher/src/main/java/org/apache/spark/launcher/JavaModuleOptions.java#L29
+exec java ${JAVA_OPTS} -Djava.io.tmpdir=/tmp \
```
Comment (reta) [inline docker/bin/start-zipkin-dependencies:13]: 👍

**Review by reta — APPROVED** [review]

**Comment by codefromthecrypt** [comment]
thanks for the look folks!

**Pull request merged**